### ISSUE TITLE
something about re-implementing rebel with no-limit heads-up Texas Hold'em

### ISSUE THREAD

**Issue by zhaoyin214** [facebookresearch/rebel#25]
I am a big fan of your excellent work and trying to re-implement rebel on no-limit heads-up Texas Hold'em.

However, I can not afford such a colossal computation of 90 DGX's, in fact, only 2 GPU's and some servers with multi-core CPU's are available. Therefore, I adopt a master-slave framework. The server with 2 GPU's is used as the master. A main process running on the master is in charge of training nets, listening on value & policy nets queries and receiving training examples from CFR processes. CFR processes are running both on the master and slaves. The communication between processes is established on ZMQ.

Also, I decreased the hyper parameters, such as, 51200 examples in each epoch (one fiftieth compared to the paper reported), limiting the subgame max-depth upto 4, 128 iterations for CFR and so on. So far, I have re-written NLHU and value net training in python (no policy net or warm up).

After training several epoches, I encountered something puzzled. The value net was not convergent. The training avg. error was increasing rapidly epoch by epoch, however, the val. avg. error was always decreasing on most current validation snapshots but increasing on much earlier snapshots.

I argue that divergency is not caused by network training with a large learning rate, but the CFR has not converged for the limited iterations. And that's the merit of the policy net. But I have no idea how to verify my argument.

Can you give me some advice? I appreciate that.

Best wishes.

Attachments: training at epoch 21, 23, 25

* epoch 25

![image](https://user-images.githubusercontent.com/42602436/136892224-b472d356-8185-4fa0-8e5f-8dfeadf87c55.png)

* epoch 23

![image](https://user-images.githubusercontent.com/42602436/136892347-fe316c58-19bf-4d18-b038-209073f0634d.png)

* epoch 21

![image](https://user-images.githubusercontent.com/42602436/136892411-c64b442a-c65a-41c1-9869-0795e6cd991e.png)

**Comment by zzzjoey** [comment]
Hi, I am also very interested in this work. I wonder how to migrate this open source code to HUNL. Have you made any progress in this area? Can we talk about it in depth? Thanks~
